### ISSUE TITLE
Support additional join types with join_where

### ISSUE THREAD

**Issue by adamreeve** [pola-rs/polars#18669]
### Description

The `join_where` method added in #18365 behaves like an inner join, where only rows from the left and right DataFrames that match the condition are returned.

It would be useful to support other join types like left, full, anti and semi.

I'm not sure how feasible it is to implement this, it seems like it would be straightforward when the join resolves to a single equi join, or single iejoin, but handling additional filter conditions could be tricky.

**Comment by ritchie46** [comment]
For equality predicates this is fairly trivial. Choose the equality predicates as outer join and add the remaining as filters. For inequality I am not entirely sure how to interpret that yet. :thinking: 

I think the first goal is to support the single inequality predicate case, as that's low hanging fruit atm!

**Comment by AlexeyDmitriev** [comment]
I think interpretation should be the same as with ON statement in SQL:

You return all the pair of rows that match all the condiitions in on (condition is calculated in terms of result)

But for left/join/outer join you'll return also rows from left/right/both table that don't have a pair

UPD: also I think it's OK if most of the conditions are not optimized for, you can start optimizing specific things and not change this interface

**Comment by adamreeve** [comment]
> I think the first goal is to support the single inequality predicate case, as that's low hanging fruit atm!

Yep I've made a start on this, I should hopefully have something to show soon

**Comment by adamreeve** [comment]
I've started looking into this and have pushed a commit with my idea of what the Python API and behaviour should look like: https://github.com/adamreeve/polars/commit/7b35d0984e73a2d16aa0d3bfcd4fa714cf13b3d0

Actually implementing this looks like it could be a fair bit of work though. For a join on equality conditions only, we could trivially convert this to an inner, left, right, full, semi or outer join. And for a join using one or two inequalities, we could fairly easily update the way join results are materialised in the IEJoin implementation to handle the different join types. But any joins that end up with extra conditions in `remaining_preds` in `resolve_join_where` would be trickier to handle.

It looks like rather than adding an extra post-processing node after the join, the best way forward would be to add support for extra non-equality predicates to all of the join implementations in order to handle this correctly and efficiently.

Consider using the existing approach of a post-processing filter for left joins for example. If we do a left join or a left IE-join and then want to apply extra join predicates in a post-processing step, we could replace values in RHS table columns with nulls where any of the new conditions are false, but would end up with extra result rows for one LHS table row if it matches with multiple RHS rows and some are filtered out in post-processing. We could probably work around that by adding a temporary row index column, but this would get pretty clunky. Semi and anti joins would be more of a problem, we'd need to do an inner join first to keep enough information to apply the extra predicates, making this a lot less efficient than it should be.

**Comment by ritchie46** [comment]
@adamreeve would it be more sensical to first implement a proper vectorized nested loop join? E.g. instead of a cross join + full materialization followed by filters we do a cartesian product on small batches + filters. This would be hit for predicates that cannot be used by IEJoin.

I feel that we can reuse some of that logic here.

**Comment by adamreeve** [comment]
Yeah that would make a lot of sense, we'd need that to handle left/full/semi/anti joins that don't use any equality or inequality conditions too as those couldn't be easily handled with a cross join and filter.

**Comment by tnk-dev** [comment]
Hey guys,
sorry for crashing this issue. 
As a user this would be the most minimalistic code snippet that should work.

```python
import polars as pl

df_1 = pl.LazyFrame({"a": [1, 2, 3], "b": [4, 5, 6]})
df_2 = pl.LazyFrame({"c": [1, 2, 3], "d": [4, 5, 6]})

df_1.join(
    df_2,
    on=(
        ((pl.col('a') == pl.lit(1)) & (pl.col('c') == pl.lit(1)))
        | (pl.col('a') == pl.lit(2))
    ),
    how="left"
).collect() # <-- throws error Column c not found

expected = pl.DataFrame(
    {
        "a": [1, 2, 3],
        "b": [4, 5, 6],
        "c": [1, 2, None],
        "d": [4, 5, None],
    }
)
```

Of course real life conditions are way more complex and nested

**Comment by EnricoMi** [comment]
Prerequisite for this is to remove `IEjoin` from `JoinType` enum. Currently, a join can either be `Inner`, `Left`, ..., *or* `IEJoin` (in-equality join). But the in-equality should be a property of the join conditions, not the join type.

So before left, semi, anti, ... joins with in-equality conditions can be supported, some refactoring will be needed to make it an ordinary (any other) join but with inequality operators (where currently any other join only allows for equality operator `==`). For this, add explicit operators to `JoinArgs`, which are used to join `left_on` against `right_on`. For current non-`IEJoin` joins, that operator is simply `==`.

Use in `JoinArgs`:
```diff
 pub struct JoinArgs {
     pub how: JoinType,
+    pub operators: Vec<JoinOperator>,
     pub validation: JoinValidation,
     pub suffix: Option<PlSmallStr>,
     pub slice: Option<(i64, usize)>,
     pub nulls_equal: bool,
     pub coalesce: JoinCoalesce,
     pub maintain_order: MaintainOrderJoin,
 }
```

Populate operators with equality (`==`) operators for each expression in `left_on` / `right_on`.
For `join_where`, extract `left_on`, `operators`, and `right_on` from `predicates`.

This makes `predicates` redundant in `Join` DSL plan:

```diff
     Join {
         input_left: Arc<DslPlan>,
         input_right: Arc<DslPlan>,
         // Invariant: left_on and right_on are equal length.
         left_on: Vec<Expr>,
         right_on: Vec<Expr>,
-        // Invariant: Either left_on/right_on or predicates is set (non-empty).
-        predicates: Vec<Expr>,
         options: Arc<JoinOptions>,
    },
```

Benefits:
- Provides the operators for `left_on` and `right_on`, currently `==` or inequality operators.
- Could replace `nulls_equal` with a null-equal operator in `operators`.
- Would allow to have different null-semantics for different join columns.
- Would allow for non-inner joins with mixture of equality and inequality operators.
- Would use current `IEJoin` implementation if `how == "inner"` and all operators have `is_inequality == true`.
- Makes `IEJoinOptions` redundant

```diff
 pub enum JoinTypeOptions {
-    #[cfg(feature = "iejoin")]
-    IEJoin(IEJoinOptions),
     Cross(CrossJoinOptions),
 }
```

Use `is_inequality` of `JoinArgs.operators` to determine if inequality operators exist, then use `IEJoin`.

Repurpose `InequalityOperator` for `JoinOperator`:
```diff
-pub enum InequalityOperator {
+pub enum JoinOperator {
     #[default]
+    Eq,
     Lt,
     LtEq,
     Gt,
     GtEq,
 }
 
-impl InequalityOperator {
+impl JoinOperator {
+    pub fn is_equality(&self) -> bool {
+        matches!(self, JoinOperator::Eq)
+    }
+
+    pub fn is_inequality(&self) -> bool {
+        !self.is_equality()
+    }
+
     pub fn is_strict(&self) -> bool {
         matches!(self, JoinOperator::Gt | JoinOperator::Lt)
     }
 }
```

Would require to integrate `resolve_join_where` into `resolve_join`.

This all is just a first step to widen expressiveness of join conditions. Next step would be to implement mixed-equality joins.

**Comment by ritchie46** [comment]
@orlp @dsprenkels, when we start with streaming non-equi joins, let's take this thread into consideration.

**Comment by adamreeve** [comment]
Looks like left join support was added in #28880, thanks!
